packages/lesswrong/server/resolvers/ultraFeedResolver.ts

```diff
@@ -0,0 +1,356 @@
+};
+
+type UsedFeedItemSourceType = Extract<FeedItemSourceType, "postThreads" | "commentThreads" | "spotlights">;
+interface WeightedSource {
+  weight: number;
+  items: FeedItem[];
+  renderAsType: FeedItemRenderType;
+}
+
+type SampledItem = { renderAsType: "feedCommentThread", feedCommentThread: FeedCommentsThread }
+                 | { renderAsType: "feedPost", feedPost: FeedFullPost }
+                 | { renderAsType: "feedSpotlight", feedSpotlight: FeedSpotlight };
+
+function weightedSample(
+  inputs: Record<UsedFeedItemSourceType, WeightedSource>, 
+  totalItems: number
+): SampledItem[] {
+  // Create deep copies of the input arrays to avoid modifying the originals
+  const sourcesWithCopiedItems = Object.entries(inputs).reduce((acc, [key, value]) => {
+    acc[key as UsedFeedItemSourceType] = {
+      ...value,
+      items: [...value.items] // Create a copy of the items array
+    };
+    return acc;
+  }, {} as Record<UsedFeedItemSourceType, WeightedSource>);
+
+  const finalFeed: SampledItem[] = [];
+  let totalWeight = Object.values(sourcesWithCopiedItems).reduce(
+    (sum, src) => sum + (src.items.length > 0 ? src.weight : 0),
+    0
+  );
+
+  for (let i = 0; i < totalItems; i++) {
+    if (totalWeight <= 0) break;
+
+    const pick = Math.random() * totalWeight;
+
+    let cumulative = 0;
+    let chosenSourceKey: UsedFeedItemSourceType | null = null;
+
+    for (const [key, src] of Object.entries(sourcesWithCopiedItems)) {
+      if (src.items.length === 0) continue;
+
+      cumulative += src.weight;
+      if (pick < cumulative) {
+        chosenSourceKey = key as UsedFeedItemSourceType;
+        break;
+      }
+    }
+
+    if (chosenSourceKey) {
+      const sourceItems = sourcesWithCopiedItems[chosenSourceKey];
+      const item = sourceItems.items.shift();
+
+      if (!item) continue;
+
+      if (sourceItems.renderAsType === "feedCommentThread") {
+        finalFeed.push({
+          renderAsType: "feedCommentThread",
+          feedCommentThread: item as FeedCommentsThread
+        });
+      } else if (sourceItems.renderAsType === "feedPost") {
+        finalFeed.push({
+          renderAsType: "feedPost",
+          feedPost: item as FeedFullPost
+        });
+      } else if (sourceItems.renderAsType === "feedSpotlight") {
+        finalFeed.push({
+          renderAsType: "feedSpotlight",
+          feedSpotlight: item as FeedSpotlight
+        });
+      }
+
+      if (sourceItems.items.length === 0) {
+        totalWeight -= sourceItems.weight;
+      }
+    }
+  }
+
+  return finalFeed;
+}
+
+
+/**
+ * UltraFeed resolver
+ */
+export const ultraFeedGraphQLQueries = {
+  UltraFeed: async (_root: void, args: any, context: ResolverContext) => {
+    const {limit = 20, cutoff, offset, sessionId} = args;
+    
+    const { currentUser } = context;
+    if (!currentUser) {
+      throw new Error("Must be logged in to fetch UltraFeed.");
+    }
+
+    try {
+      const ultraFeedRepo = new UltraFeedRepo();
```

Comment:
you can pull this from the resolver context rather than importing and creating a new instance

Reply:
(you can also pull the collection out of the context rather than importing it)